src/thorin/pass/rw/lower_for.cpp

```diff
@@ -0,0 +1,54 @@
+#include "thorin/pass/rw/lower_for.h"
+
+#include "thorin/lam.h"
+
+namespace thorin {
+
+const Def* LowerFor::rewrite(const Def* def) {
+    if (auto app = def->isa<App>())
+        if (auto for_ax = app->axiom(); for_ax && for_ax->tag() == Tag::For) {
+            auto& w = def->world();
+            w.DLOG("rewriting for axiom: {} within {}", app, curr_nom());
+
+            auto for_pi  = app->callee_type();
+            auto for_lam = w.nom_lam(for_pi, w.dbg("for"));
+
+            auto body      = app->args()[app->num_args() - 2];
```

Comment:
see comment below. Maybe you want to use a destructering pattern.

Reply:
I find that too spilly as I only care once for the body and the other args are just all reused as a batch right now.